keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,559 @@
+
+2. **Variable Naming**:
+    
+    a. **Standard**: Environment variable names should be composed of alphanumeric characters (letters and numbers), underscores, dots, or hyphens. The first character cannot be a digit.
+
+    b. **Relaxed**: If the Kubernetes feature gate `RelaxedEnvironmentVariableValidation` is enabled, the naming restriction is less stringent, allowing any printable ASCII character except the equals sign (=).
+
+3. **Duplicate Names**: If an environment variable is defined multiple times in the file, the last occurrence takes precedence and overrides any previous values.
+
+4. **Size Limit**: The maximum allowed size for the env file is 1 MiB.
+
+5. **File Location**: The env file must be placed within the `emptyDir` volume associated with the pod. If it is not found in the correct location, the Kubernetes API server will reject the pod creation request.
+
+6. **Container Behavior**:
+    
+    a. **Startup**: At container startup, the kubelet (the Kubernetes node agent) will parse the env file from the emptyDir volume and inject the defined variables into the container's environment.
+    
+    b. **File Access**: The env file itself is not directly accessible within the container unless explicitly mounted by the container configuration.
+
+7. **Dynamic Updates**: Currently, changes to the env file after container startup are not reflected in the container's environment until a restart. A potential future enhancement could involve dynamically updating environment variables when the file is modified.
+
+
+### Failure and Fallback Strategy
+
+There are different scenarios in which applying `FileEnvSource` and
+`FileKeySelector` fields may fail. We plan to provide an erorr message that is
+descriptive enough to troubleshoot the issue but it will not contain the env
+file as it can contain sensitive information.
+
+Below are the ones we mapped and their outcome once this KEP is implemented.
+
+|Scenario| API Server Result | Kubelet Result |
+|--------|-------------------|----------------|
+|1. The volumeName specified in `FileEnvSource` or `FileKeySelector` field does not exist | Pod creation fails with an error |  |
+|2. The volumeName specified in `FileEnvSource` or `FileKeySelector` is not an emptyDir | Pod creation fails with an error |  |
+|3. The filepath specified in `FileEnvSource` field does not exist | Pod created | Container fails to start and error message in event.|
+|4. The filepath specified in `FileEnvSource` field does not exist but `optional` field is set to true. | Pod created | Container starts and env vars are not populated.|
+|5. Either the filepath or key specified in `FileKeySelector` field does not exist | Pod created | Container fails to start and error message in event.|
+|6. Either the filepath or key specified in `FileKeySelector` field exist but `optional` field is set to true | Pod created | Container starts and env vars are not populated. |
```

Comment:
should this be `does not exist` ?
if the file doesnt exist, and optional is true - shouldn't we error out here?

Reply:
Fixed.